File: 2024/24/solution.py
from functools import cache, cached_property
from itertools import permutations
init,prog = open(0).read().split("\n\n")

regs = {}

for reg in init.splitlines():
    name,val = reg.split(": ") 
    regs[name] = int(val)

for gates in prog.strip().splitlines():
    inst, name  = gates.split(" -> ")
    regs[name] = inst   

@cache
def calc(reg):
    val = regs[reg]
    if val == 1 or val == 0: return val
    inst = regs[reg]
    p1,op,p2 = inst.split(" ")
    v1 = calc(p1)
    v2 = calc(p2)
    result = None
    if op == "AND":
        result = v1 & v2
    elif op == "OR":
        result = v1 | v2
    elif op == "XOR":
        result = v1 ^ v2
    else:
        raise RuntimeError("Invalid op:", op)
    
    # regs[reg] = result
    return result 


z_regs = [k for k in regs.keys() if k.startswith("z")]
x_regs = [k for k in regs.keys() if k.startswith("x")]
y_regs = [k for k in regs.keys() if k.startswith("y")]

x = int("".join(str(calc(name)) for name in (reversed(sorted(x_regs)))),2)
y = int("".join(str(calc(name)) for name in (reversed(sorted(y_regs)))),2)
z = int("".join(str(calc(name)) for name in (reversed(sorted(z_regs)))),2)

print("part1:" ,z)
print(x+y)

print(len(regs.keys()))
print(312*311/2*310*309/2*308*307/2*306*305/2)
# print(len(list(permutations(regs.keys(),8))))

File: 2024/24/solution.rb
init,program = $stdin.read.split("\n\n")

regs = {}
init.lines.each { |l|
  name,val = l.strip.split(": ")
  regs[name] = val.to_i
}

program.lines.each { | l |
  inst,reg = l.strip.split(" -> ")
  regs[reg] = inst
}


def calc(reg,regs)
  val = regs[reg]
  # p ["v", val]
  if val == 0 || val == 1
    return val 
  end
  p1, op , p2 = val.split
  # p [p1,p2]
  p1 = calc(p1,regs)
  p2 = calc(p2,regs)
  # p [op, p1,p2]
  case (op)
  when "AND"
    return p1 & p2
  when "OR"
    return p1 | p2
  when "XOR"
    return (p1 ^ p2)
  end 
end

p regs.keys.filter { |it| it.start_with?('z')}.sort.reverse.map { |it| calc(it,regs) }.join("").to_i(2)
